Chisel/CHLFunctions.m

```diff
@@ -0,0 +1,65 @@
+//
+//  CHLFunctions.m
+//  Chisel
+//
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "CHLFunctions.h"
+
+// See vm_range_recorder_t
+static void enumerateRanges(task_t, void *, unsigned, vm_range_t *, unsigned);
+
+void CHLEnumerateAllocationsWithBlock(void (^block)(vm_range_t))
+{
+  /*
+   * === Overview ===
+   *
+   * To enumerate the set of live memory allocations ("ranges"), the pseudocode might look like:
+   *
+   *     for zone in all_zones()
+   *       for range in zone->ranges
+   *         callback(range)
+   *
+   * Unfortunately, the memory allocations within a zone are not exposed, so the above inner loop is not possible.
+   * Instead malloc zones provide a function that will loop over the allocations and call the given callback function.
+   */
+
+  /*
+   * === malloc API Notes ===
+   *
+   * Both `malloc_get_all_zones` and `vm_range_recorder_t` callback functions take a `task_t` and a `memory_reader_t`.
+   * However it turns out both of these can be NULL. This information isn't documented in malloc.h but can be seen in
+   * the libmalloc source.
+   *
+   * `memory_reader_t`:
+   * When NULL, the default memory reader is used which handles in process memory, and does not use the given `task_t`.
+   * `task_t`:
+   * Only required if used in a custom `memory_reader_t`, or used in a `vm_range_recorder_t` callback.
+   */
+
+  // Get a reference to the C array of malloc zones.
+  vm_address_t *zones;
+  unsigned int zoneCount;
+  malloc_get_all_zones(TASK_NULL, NULL, &zones, &zoneCount);
+
+  // Outer loop over the zones.
+  for (unsigned int i = 0; i < zoneCount; i++) {
+    // While `zones` is typed as `vm_address_t *` for type checking purposes, these addresses are just pointers to
+    // `malloc_zone_t` structs. Thus `vm_address_t` and `malloc_zone_t *` are used interchangably here.
+    malloc_zone_t *zone = (malloc_zone_t *)zones[i];
+
+    // Inner loop over the zone's memory allocations.
+    zone->introspect->enumerator(TASK_NULL, block, MALLOC_PTR_IN_USE_RANGE_TYPE, (vm_address_t)zone, NULL, enumerateRanges);
```

Comment:
Theoretically yes, in practice it's currently always present. I wanted to try to make this code easy to digest, so I left out the NULL checks here and also don't check the return value of `malloc_get_all_zones`. In part I feel semi justified because this is intended for production use, just debugging. Thoughts?